pallets/data-preservers/src/lib.rs

```diff
@@ -550,22 +612,133 @@ pub mod pallet {
+
+            let witness = T::AssignmentPayment::try_start_assignment(
+                assigner,
+                profile.account.clone(),
+                &profile.profile.assignment_request,
+                assigner_param,
+            )?;
+
+            profile.assignment = Some((para_id, witness));
+            Profiles::<T>::insert(profile_id, profile);
+
+            Self::deposit_event(Event::AssignmentStarted {
+                profile_id,
+                para_id,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(8)]
+        #[pallet::weight(0)]
+        pub fn stop_assignment(
+            origin: OriginFor<T>,
+            profile_id: T::ProfileId,
+            para_id: ParaId,
+        ) -> DispatchResultWithPostInfo {
+            let caller =
+                EitherOfDiverse::<T::AssignmentOrigin, EnsureSigned<T::AccountId>>::ensure_origin(
+                    origin, &para_id,
+                )?;
+
+            let mut profile = Profiles::<T>::get(profile_id).ok_or(Error::<T>::UnknownProfileId)?;
+
+            match caller {
+                // para id manager, is allowed to call
+                Either::Left(_) => (),
+                // signed, must be profile owner
+                Either::Right(account) => ensure!(
+                    profile.account == account,
+                    sp_runtime::DispatchError::BadOrigin
+                ),
+            }
+
+            let Some((assignment_para_id, assignment_witness)) = profile.assignment.take() else {
+                Err(Error::<T>::ProfileNotAssigned)?
+            };
+
+            if assignment_para_id != para_id {
+                Err(Error::<T>::WrongParaId)?
+            }
+
+            T::AssignmentPayment::try_stop_assignment(profile.account.clone(), assignment_witness)?;
+
+            Profiles::<T>::insert(profile_id, profile);
+
+            {
+                let mut assignments = Assignments::<T>::get(para_id);
+                if !assignments.remove(&profile_id) {
+                    Err(Error::<T>::ProfileNotAssigned)?
+                }
+                Assignments::<T>::insert(para_id, assignments);
+            }
+
+            Self::deposit_event(Event::AssignmentStopped {
+                profile_id,
+                para_id,
+            });
+
+            Ok(().into())
+        }
     }
 
     impl<T: Config> Pallet<T> {
+        pub fn assignments_profiles(para_id: ParaId) -> impl Iterator<Item = Profile<T>> {
+            Assignments::<T>::get(para_id)
+                .into_iter()
+                .filter_map(Profiles::<T>::get)
+                .map(|profile| profile.profile)
+        }
+
         /// Function that will be called when a container chain is deregistered. Cleans up all the storage related to this para_id.
         /// Cannot fail.
         pub fn para_deregistered(para_id: ParaId) {
-            BootNodes::<T>::remove(para_id);
+            Assignments::<T>::remove(para_id);
```

Comment:
This however doesn't remove assignment in each profile.

Reply:
this is a good question in fact. `para_deregistered` can only happen in session changes right? here we use max_weight but depending on the amount of deregistration this can get quite messy

Because otherwise what is the alternative? list in the dapp anything that is either free or assigned to a de-registerd chain?

Reply:
I guess we could also make a non-paid extrinsic to clean it up..

Reply:
>para_deregistered can only happen in session changes right?

Almost always, there is an edge case when deregistering a pending chain, which is immediate, not sure if we allow pending chains to start an assignment... I guess yes because they should make sure to have bootnodes before calling `mark_valid_for_collating`.

> Because otherwise what is the alternative?

Having some cleanup extrinsic, or doing it using on_idle. I would probably delete the first 1 or 2 assigned profiles automatically, which should handle most use cases, and then use the lazy unassignment for the rest.

Reply:
the on_idle case is actually not a bad idea...but only if we have a onIdleUnassigment storage item where we push profile-ids